examples/kitchen-sink.yaml

```diff
@@ -12,358 +12,404 @@ file_format: "0.1"
-            protocol: http/protobuf
-            # Configure endpoint.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_ENDPOINT, OTEL_EXPORTER_OTLP_LOGS_ENDPOINT
-            endpoint: http://localhost:4318
-            # Configure certificate.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_CERTIFICATE, OTEL_EXPORTER_OTLP_LOGS_CERTIFICATE
-            certificate: /app/cert.pem
-            # Configure mTLS private client key.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_CLIENT_KEY, OTEL_EXPORTER_OTLP_LOGS_CLIENT_KEY
-            client_key: /app/cert.pem
-            # Configure mTLS client certificate.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_CLIENT_CERTIFICATE, OTEL_EXPORTER_OTLP_LOGS_CLIENT_CERTIFICATE
-            client_certificate: /app/cert.pem
-            # Configure headers.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_HEADERS, OTEL_EXPORTER_OTLP_LOGS_HEADERS
-            headers:
-              api-key: "1234"
-            # Configure compression.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_COMPRESSION, OTEL_EXPORTER_OTLP_LOGS_COMPRESSION
-            compression: gzip
-            # Configure max time (in milliseconds) to wait for each export.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_TIMEOUT, OTEL_EXPORTER_OTLP_LOGS_TIMEOUT
-            timeout: 10000
-            # Configure client transport security for the exporter's connection.
-            #
-            # Environment variable: OTEL_EXPORTER_OTLP_INSECURE, OTEL_EXPORTER_OTLP_LOGS_INSECURE
-            insecure: false
+    batch:
+      # Configure delay interval (in milliseconds) between two consecutive exports.
+      #
+      # Environment variable: OTEL_CONF_LOGGER_PROVIDER__PROCESSORS__BATCH__SCHEDULE_DELAY
+      schedule_delay: 5000
+      # Configure maximum allowed time (in milliseconds) to export data.
+      #
+      # Environment variable: OTEL_CONF_LOGGER_PROVIDER__PROCESSORS__BATCH__EXPORT_TIMEOUT
+      export_timeout: 30000
+      # Configure maximum queue size.
+      #
+      # Environment variable: OTEL_CONF_LOGGER_PROVIDER__PROCESSORS__BATCH__MAX_QUEUE_SIZE
+      max_queue_size: 2048
+      # Configure maximum batch size.
+      #
+      # Environment variable: OTEL_CONF_LOGGER_PROVIDER__PROCESSORS__BATCH__MAX_EXPORT_BATCH_SIZE
+      max_export_batch_size: 512
+      # Configure exporter.
+      exporter:
+        # Configure exporter to be OTLP.
+        otlp:
+          # Configure protocol.
+          #
+          # Environment variable: OTEL_CONF_LOGGER_PROVIDER__PROCESSORS__BATCH__EXPORTER__OTLP__PROTOCOL
+          protocol: http/protobuf
+          # Configure endpoint.
+          #
+          # Environment variable: OTEL_CONF_LOGGER_PROVIDER__PROCESSORS__BATCH__EXPORTER__OTLP__ENDPOINT
```

Comment:
> I understand the ambiguity that brings tho...

To make this concrete.. 

Suppose we want to change the `max_queue_size` of the batch span processor given the following config
```
tracer_provider:
  processors:
    batch:
      exporter:
        otlp:
          endpoint: http://endpoint1
  processor_pipeline: [batch]
```

If we use spring's mappings, we'd set something like:
```
OTEL_CONF_TRACERPROVIDER_PROCESSORS_BATCH_MAXQUEUESIZE=10
```

In order to apply this override, we have to know the structure of the schema ahead of time. If we don't, and just blindly try to update the raw YAML, we'd get:
```
tracer_provider:
  processors:
    batch:
      exporter:
        otlp:
          endpoint: http://endpoint1
  processor_pipeline: [batch]
tracerprovider:
  processors:
    batch:
      maxqueuesize: 10
```
In this case, we know that the batch processor has a property called `max_queue_size`. So when evaluating, we can take the position of this property `.tracer_provider.processors.batch.max_queue_size`, apply a generic transformation yielding `OTEL_CONF_TRACERPROVIDER_PROCESSORS_BATCH_MAXQUEUESIZE` and check if any environment variable is set which matches.

But we don't know the schema ahead of time for custom components. Consider a custom span exporter named `foo` with a property called `my_property`:
```
tracer_provider:
  processors:
    batch:
      exporter:
        foo:
          my_property: http://endpoint1
```
Because we don't know the schema of the foo exporter ahead of time, we have to way to tell that `OTEL_TRACERPROVIDER_PROCESSORS_BATCH_EXPORTER_FOO_MYPROPERTY` should target `my_property` or `myproperty`. 


Prior art that uses the `__` include python's [dynaconf](https://dynaconf.readthedocs.io/en/docs_223/guides/environment_variables.html#nested-keys-in-dictionaries-via-environment-variables) and [.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0#non-prefixed-environment-variables).

Reply:
> Prior art that uses the __ include python's [dynaconf](https://dynaconf.readthedocs.io/en/docs_223/guides/environment_variables.html#nested-keys-in-dictionaries-via-environment-variables) and [.NET Core](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/configuration/?view=aspnetcore-8.0#non-prefixed-environment-variables).

👍